### ISSUE TITLE
Optional unicode nickname support

### ISSUE THREAD

**Issue by simnalamburt** [martynsmith/node-irc#456]
#### TL;DR

Hey it's 2016 and it's time to use unicode.

--------

Most IRC clients allows unicode nickname; irssi, weechat, MozIRC, IRCCloud, shoutIRC, ZNC, etc. And most CJK IRC servers allows unicode nickname; http://ozinger.org/, http://www.hanirc.org/, http://www.luatic.net/, https://twitter.com/HanIRC, etc. I know that [current IRC spec][spec] does not allow non-ASCII nickname but that spec has not been changed for 16 years and CJK and non American users **cannot** use IRC without unicode nickname support.

[spec]: https://tools.ietf.org/html/rfc2812#section-2.3.1

![2016-03-26 4 53 40](https://cloud.githubusercontent.com/assets/4435445/14052591/bb711958-f30e-11e5-82c2-f9136a305ac3.png)

![](https://cloud.githubusercontent.com/assets/4435445/14052862/9f418180-f310-11e5-8eb7-3ab95fa5a9bc.png)

And there were so many requests that asks node-irc to allow non-ASCII nickname. And there're so many forks which enables non-ASCII nickname. It's 2016 and it's a shame to know that node-irc does not support non-ASCII nickname while everyone else does.

- https://github.com/martynsmith/node-irc/pull/377, https://github.com/martynsmith/node-irc/pull/377/commits/78847c13b3c133aacfb90c5314b564aaf296053e (@outsideris, @kimdwkimdw)
- https://github.com/martynsmith/node-irc/pull/408, https://github.com/martynsmith/node-irc/pull/408/commits/31b790618fb817fc765cfe60b10ed576beced1d9 (@leeopop)
- https://github.com/VBChunguk/node-irc/commit/a2f02f76559660de897333e35b5cbe7c72afa79a (@VBChunguk)
- https://github.com/martynsmith/node-irc/issues/104 (@sinoru)
- https://github.com/martynsmith/node-irc/pull/175 (@ejnahc)

**Comment by jirwin** [comment]
Thanks for this report. I'll take a look at this today and see what I can do.

**Comment by Trinitas** [comment]
Send/write nick as binary to the server?
('NICK '+nick+'\r\n', 'binary');

**Comment by Throne3d** [comment]
I'm not particularly well-acquainted with this project, but I'm not actually sure what spec we should be going by that permits Unicode nicknames? I haven't seen a specific one mentioned, looking around the project briefly, but [RFC2812](https://tools.ietf.org/html/rfc2812#section-2.3.1) doesn't allow them (as said in the start of this issue):
```
message    =  [ ":" prefix SPACE ] command [ params ] crlf
prefix     =  servername / ( nickname [ [ "!" user ] "@" host ] )
nickname   =  ( letter / special ) *8( letter / digit / special / "-" )
letter     =  %x41-5A / %x61-7A       ; A-Z / a-z
digit      =  %x30-39                 ; 0-9
special    =  %x5B-60 / %x7B-7D
                 ; "[", "]", "\", "`", "_", "^", "{", "|", "}"
```

At present it looks like the code matches this spec (mostly) – it looks in the prefix for the Regex pattern:
```
/^([_a-zA-Z0-9\~\[\]\\`^{}|-]*)(!([^@]+)@(.*))?$/
```
(the 1st, 3rd and 4th matches are made into the `nick`, `user` and `host` params, or if nothing is matched, it makes the whole prefix into the message's `server`)

I also found [this spec](http://modern.ircdocs.horse/) which is apparently in the process of being written and seems to be part of '[IRCv3](http://ircv3.net/)', but it doesn't yet specify what can go into a nickname.

Any ideas?

**Edit**: Or we could just hope to merge one of the already existing PRs, I guess.

**Comment by dagenim** [comment]
sorry I can not speak english well.

I had this problem with teleirc.

<pre><code>match = message.prefix.match(/^([_a-zA-Z0-9\~\[\]\\`^{}|-]*)(!([^@]+)@(.*))?$/);</code></pre>

this line edit to

<pre><code>match = message.prefix.match(/^([_a-zA-Z0-9\~\[\]\\`^{}|-|\u0100-\ud800]*)(!([^@]+)@(.*))?$/);</code></pre>

and work well with korean
